File: src/main/java/io/github/vehkiya/service/provider/JsonDataProvider.java
package io.github.vehkiya.service.provider;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.github.vehkiya.config.ServiceProviderProperties;
import io.github.vehkiya.data.model.domain.Item;
import io.github.vehkiya.data.model.json.JsonItem;
import io.github.vehkiya.data.model.json.JsonSource;
import io.github.vehkiya.service.DataProvider;
import lombok.Getter;
import lombok.experimental.Accessors;
import org.springframework.beans.factory.annotation.Autowired;

import javax.annotation.PostConstruct;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;

public class JsonDataProvider implements DataProvider {

    @Autowired
    private ServiceProviderProperties serviceProviderProperties;

    @Getter
    @Accessors(fluent = true)
    private Map<String, Item> itemsCache;

    @Autowired
    private ObjectMapper objectMapper;

    @PostConstruct
    public void init() throws IOException {
        itemsCache = new ConcurrentHashMap<>();
        refresh();
    }

    @Override
    public void refresh() throws IOException {
        itemsCache.clear();
        var jsonSource = readJsonFile();
        jsonSource.getItems().values()
                .stream()
                .filter(Objects::nonNull)
                .filter(jsonItem -> Objects.nonNull(jsonItem.name()))
                .map(itemConverter())
                .forEach(item -> itemsCache.put(item.name(), item));
    }

    private Function<JsonItem, Item> itemConverter() {
        return jsonItem -> new Item().name(jsonItem.name());
    }

    private JsonSource readJsonFile() throws IOException {
        var path = Paths.get(serviceProviderProperties.getSource());
        if (!path.toFile().exists()) {
            throw new IllegalArgumentException("Source file " + path + " not found ");
        }
        var content = new String(Files.readAllBytes(path));
        return objectMapper.readValue(content, JsonSource.class);
    }


}

File: src/main/java/io/github/vehkiya/service/provider/JsonDataProvider.kt
package io.github.vehkiya.service.provider

import com.fasterxml.jackson.databind.ObjectMapper
import io.github.vehkiya.config.ServiceProviderProperties
import io.github.vehkiya.data.model.domain.Item
import io.github.vehkiya.data.model.json.JsonSource
import io.github.vehkiya.service.DataProvider
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.nio.file.Files
import kotlin.io.path.Path
import kotlin.io.path.notExists

@Service
class JsonDataProvider
@Autowired constructor(
    val objectMapper: ObjectMapper,
    val properties: ServiceProviderProperties
) : DataProvider {

    private val itemCache: MutableMap<String, Item> = mutableMapOf()

    init {
        refresh()
    }

    override fun itemCache(): Map<String, Item> {
        return itemCache
    }

    final override fun refresh() {
        itemCache.clear()
        val jsonSource = readJsonFile()
        jsonSource.items
            .values
            .map { convert { Item(it.name) } }
            .forEach { itemCache[it.name] = it }
    }

    private inline fun <T> convert(mapper: () -> T): T = mapper.invoke()

    private fun readJsonFile(): JsonSource {
        val path = Path(properties.source)
        if (path.notExists()) {
            throw IllegalArgumentException("Source file $path not found")
        }
        val content = Files.readAllBytes(path)
        return objectMapper.readValue(content, JsonSource::class.java)
    }
}